### PR TITLE
Rimpl depth

### PR DESCRIPTION
**Pull request by narekhovhannisyan**
### Types of changes

- [x] Enhancement (project structure, spelling, grammar, formatting)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### A description of the changes proposed in the Pull Request

- Rimpl script now supports nested children.

### CODE REVIEW

File: examples/impls/nnt-data-on-premise.yml [hunk 1]
```diff
@@ -0,0 +1,164 @@
+name: ntt-data-on-premise
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:1]: the filename should start `ntt` not `nnt` and it should overwrite the file already in the example/impls directory

File: examples/impls/nnt-data-on-premise.yml [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+          five-min-output-rate: 100
+          grid-ci: 951 # gCO2e/kWh
+          energy: 0.0004554559930455524
+    layer-2-switch:
+      pipeline:
+        - sci-m # duration & config -> embodied
+        - sci-o # energy & grid-carbon-intensity & embodied -> carbon
+      config:
+        sci-m:
+          te: 0.251 # kgCO2eq
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:45]: same
```suggestion
          te: 251 # gCO2eq
```

File: examples/impls/nnt-data-on-premise.yml [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+              duration: 1 #s
+              grid-ci: 951 # gCO2e/kWh
+              energy: 0.0004554559930455524
+    app-servers:
+      pipeline:
+        - sci-m # duration & config -> embodied
+        - sci-o # energy & grid-carbon-intensity & embodied -> carbon
+      config:
+        sci-m:
+          te: 6270 # kgCO2eq
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:115]:
```suggestion
          te: 6270000 # gCO2eq
```

File: examples/impls/nnt-data-on-premise.yml [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+              e: 0.147
+              grid-ci: 951 # gCO2e/kWh
+              energy: 0.0004554559930455524
+    web-servers:
+      pipeline: 
+        - sci-m # duration & config -> embodied
+        - sci-o # energy & grid-carbon-intensity & embodied -> carbon 
+      config:
+        sci-m:
+          te: 6270 # kgCO2eq
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:143]:
```suggestion
          te: 6270000 # gCO2eq
```

File: examples/impls/nnt-data-on-premise.yml [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+          rr: 1
+          tor: 1
+        sci-o:
+          i: 457 # gCO2/kwh 
+      children:
+        switch-1:
+          observations:
+            - timestamp: 2023-07-06T00:00
+              duration: 1
+              e: 5.41e-4 #kwh
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:57]: How is rimpl handling values associated to `e`  keys? In the `case-study-models` branch we expect `e_cpu` `e-mem` or `e-net` and components and `sci-e` returns `energy`, so I was surprised this impl ran. In any case, we should update to `e-net` here as it most accurately describes the energy used by the switch.
```suggestion
              e_net: 5.41e-4 #kwh     
```

File: examples/impls/nnt-data-on-premise.yml [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+            - timestamp: 2023-07-06T00:00
+              duration: 1
+              e: 5.41e-4 #kwh     
+              grid-ci: 951 # gCO2e/kWh
+              energy: 0.0004554559930455524
+        switch-2:
+          observations: 
+            - timestamp: 2023-07-06T00:00
+              duration: 1
+              e: 0 # no traffic at all
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:64]:
```suggestion
              e_net: 0 # no traffic at all
```

File: examples/impls/nnt-data-on-premise.yml [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+              grid-ci: 951 # gCO2e/kWh
+              energy: 0.0004554559930455524
+        switch-3:
+          observations:
+            - timestamp: 2023-07-06T00:00
+              duration: 1
+              e: 6.37e-4 #kwh
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:71]:
```suggestion
              e-net: 6.37e-4 #kwh
```

File: examples/impls/nnt-data-on-premise.yml [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+              grid-ci: 951 # gCO2e/kWh
+              energy: 0.0004554559930455524
+        switch-4:
+          observations:
+            - timestamp: 2023-07-06T00:00
+              duration: 1
+              e: 7.60e-9 #kwh
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:78]:
```suggestion
              e-net: 7.60e-9 #kwh
```

File: examples/impls/nnt-data-on-premise.yml [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+          el: 126144000 # 4 years in seconds
+          rr: 1
+          tor: 1
+        sci-o:
+          i: 457 # gCO2/kwh
+      children:
+        db-server1:
+          observations:
+            - timestamp: 2023-07-06T00:00
+              e: 0.258
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:98]:
```suggestion
              e-cpu: 0.258
```

File: examples/impls/nnt-data-on-premise.yml [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+              duration: 1 #s
+              grid-ci: 951 # gCO2e/kWh
+              energy: 0.0004554559930455524
+        db-server2:
+          observations:
+            - timestamp: 2023-07-06T00:00
+              e: 0.193
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:105]:
```suggestion
              e_cpu: 0.193
```

File: examples/impls/nnt-data-on-premise.yml [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+          rr: 1 
+          tor: 1 
+        sci-o:
+          i: 457 # gCO2/kwh         
+      children:
+        app-server1:
+          observations: 
+            - timestamp: 2023-07-06T00:00
+              duration: 1
+              e: 0.149
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:127]:
```suggestion
              e-cpu: 0.149
```

File: examples/impls/nnt-data-on-premise.yml [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+          rr: 1 
+          tor: 1 
+        sci-o:
+          i: 457 # gCO2/kwh  
+      children:
+        web-server1:
+          observations: 
+            - timestamp: 2023-07-06T00:00
+              duration: 1
+              e: 0.139
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:155]:
```suggestion
              e-cpu: 0.139
```

File: examples/impls/nnt-data-on-premise.yml [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+              grid-ci: 951 # gCO2e/kWh
+              energy: 0.0004554559930455524
+        web-server2:
+          observations: 
+            - timestamp: 2023-07-06T00:00
+              duration: 1
+              e: 0.116
```
Comment (jmcook1186) [inline examples/impls/nnt-data-on-premise.yml:162]:
```suggestion
              e-cpu: 0.116
```

**Review by jmcook1186 — APPROVED** [review]
sgtm - all existing tests passing locally. Would prefer to see a new test added that pulls in an impl with several levels of nesting to confirm the new rimpl functionality code behaves as expected.

Also think it needs some detailed documentation as the naming is not super intuitive (I'm really referring to `supercomputer` - I like the name but it's not immediately obvious what it does. Not blocking, but this should be added at some point.

**Comment by narekhovhannisyan** [comment]
Agree, we need to cover rimpl with tests also. Will think of different name selection which will be beautiful, but also fully reflective

**Pull request merged**